range-requests/src/lib.rs

```diff
@@ -0,0 +1,530 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use dropshot::Body;
+use dropshot::HttpError;
+use futures::TryStreamExt;
+use http::HeaderValue;
+use hyper::{
+    header::{ACCEPT_RANGES, CONTENT_LENGTH, CONTENT_RANGE, CONTENT_TYPE},
+    Response, StatusCode,
+};
+
+const ACCEPT_RANGES_BYTES: http::HeaderValue =
+    http::HeaderValue::from_static("bytes");
+const CONTENT_TYPE_OCTET_STREAM: http::HeaderValue =
+    http::HeaderValue::from_static("application/octet-stream");
+
+/// Errors which may be returned when processing range requests
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Using multiple ranges is not supported")]
+    MultipleRangesUnsupported,
+
+    #[error("Range would overflow (start + length is too large)")]
+    RangeOverflow,
+
+    #[error("Range would underflow (total content length < start)")]
+    RangeUnderflow,
+
+    #[error("Empty Range")]
+    EmptyRange,
+
+    #[error("Failed to parse range: {0:?}")]
+    Parse(http_range::HttpRangeParseError),
+
+    #[error(transparent)]
+    Http(#[from] http::Error),
+}
+
+impl From<Error> for HttpError {
+    fn from(err: Error) -> Self {
+        match err {
+            Error::MultipleRangesUnsupported => {
+                HttpError::for_bad_request(None, err.to_string())
+            }
+            Error::RangeUnderflow
+            | Error::RangeOverflow
+            | Error::EmptyRange => HttpError::for_client_error(
+                None,
+                http::StatusCode::RANGE_NOT_SATISFIABLE,
+                "Range was parseable, but invalid".to_string(),
+            ),
+            Error::Parse(_) => HttpError::for_client_error(
+                None,
+                http::StatusCode::RANGE_NOT_SATISFIABLE,
```

Comment:
Sorry to continue nitpicking about this, but I *thought* that only parseable-but-nonsensical range requests were supposed to return this status, and one that's syntactically malformed should get Bad Request instead?

Reply:
good feedback, I'll change this case to "bad request".